src/ImageFitter/ImageFitter.cc

```diff
@@ -0,0 +1,227 @@
+        SetResults();
+        SetLog();
+    }
+
+    gsl_vector_free(_fit_params);
+    gsl_vector_free(_fit_errors);
+    return success;
+}
+
+void ImageFitter::SetFitData(float* image, size_t width, size_t height) {
+    size_t n = width * height;
+    _fit_data.data.resize(n);
+    _fit_data.x.resize(n);
+    _fit_data.y.resize(n);
+
+    size_t i = 0;
+    for (size_t j = 0; j < height; j++) {
+        for (size_t k = 0; k < width; k++) {
+            double data = image[j * width + k];
+            if (!isnan(data)) {
+                _fit_data.data[i] = data;
+                _fit_data.x[i] = k;
+                _fit_data.y[i] = j;
+                i++;
+            } else {
+                n--;
+            }
+        }
+    }
+
+    _fit_data.n = n;
+    _fit_data.data.resize(n);
+    _fit_data.x.resize(n);
+    _fit_data.y.resize(n);
+    _fdf.n = n;
+}
+
+void ImageFitter::SetInitialValues(const std::vector<CARTA::GaussianComponent>& initial_values) {
+    _num_components = initial_values.size();
+
+    size_t p = _num_components * 6;
+    _fit_params = gsl_vector_alloc(p);
+    _fit_errors = gsl_vector_alloc(p);
+    for (size_t i = 0; i < _num_components; i++) {
+        CARTA::GaussianComponent component(initial_values[i]);
+        gsl_vector_set(_fit_params, i * 6, component.center_x());
+        gsl_vector_set(_fit_params, i * 6 + 1, component.center_y());
+        gsl_vector_set(_fit_params, i * 6 + 2, component.amp());
+        gsl_vector_set(_fit_params, i * 6 + 3, component.fwhm_x());
+        gsl_vector_set(_fit_params, i * 6 + 4, component.fwhm_y());
+        gsl_vector_set(_fit_params, i * 6 + 5, component.pa());
+    }
+    _fdf.p = p;
+}
+
+int ImageFitter::SolveSystem() {
+    gsl_multifit_nlinear_parameters fdf_params = gsl_multifit_nlinear_default_parameters();
+    const gsl_multifit_nlinear_type* T = gsl_multifit_nlinear_trust;
+    const size_t max_iter = 200;
+    const double xtol = 1.0e-8;
+    const double gtol = 1.0e-8;
+    const double ftol = 1.0e-8;
+    const bool print_iter = false;
+    const size_t n = _fdf.n;
+    const size_t p = _fdf.p;
+    gsl_multifit_nlinear_workspace* work = gsl_multifit_nlinear_alloc(T, &fdf_params, n, p);
+    gsl_vector* f = gsl_multifit_nlinear_residual(work);
+    gsl_vector* y = gsl_multifit_nlinear_position(work);
+    gsl_matrix* covar = gsl_matrix_alloc(p, p);
+
+    gsl_multifit_nlinear_init(_fit_params, &_fdf, work);
+    gsl_blas_ddot(f, f, &_fit_status.chisq0);
+    int status = gsl_multifit_nlinear_driver(max_iter, xtol, gtol, ftol, print_iter ? Callback : NULL, NULL, &_fit_status.info, work);
+    gsl_blas_ddot(f, f, &_fit_status.chisq);
+    gsl_multifit_nlinear_rcond(&_fit_status.rcond, work);
+    gsl_vector_memcpy(_fit_params, y);
+
+    gsl_matrix* jac = gsl_multifit_nlinear_jac(work);
+    gsl_multifit_nlinear_covar(jac, 0.0, covar);
+    const double c = GSL_MAX_DBL(1, sqrt(_fit_status.chisq / (n - p)));
```

Comment:
why do we need to use GSL_MAX_DBL to get the max of two double values? Here it is 1 and sqrt(_fit_status.chisq / (n - p)).

Reply:
sqrt(_fit_status.chisq / (n - p)) is used when estimating errors of a poor fit. Please see https://www.gnu.org/software/gsl/doc/html/nls.html#exponential-fitting-example for example.

Reply:
yes but why do we need to compare it to unity and pick up the greater one as c?

Reply:
to check if the chi-squared value shows a poor fit